### PR TITLE
PDT-618 Redo redis.py

### PR DESCRIPTION
**Pull request by nicwolff**
The previous merge of this branch failed the Docker build on Travis when it tries to look up the Redis primary's hostname – which doesn't exist there since we don't build the other containers.

Trying to fix that was made difficult by our repeated patches to the redis.py module which have made it very hard to maintain. This change simplifies the Redis client class construction, and makes the list of read replicas an attribute of that class instead of a global variable. It also cleans up the settings names.

### CODE REVIEW

File: cacheops/redis.py [hunk 1]
```diff
@@ -1,20 +1,37 @@
 from __future__ import absolute_import
-import warnings
+
+import os.path
```
Comment (nicwolff) [inline cacheops/redis.py:3]: I sorted the imports.

File: cacheops/redis.py [hunk 2]
```diff
@@ -1,20 +1,37 @@
...
 
 from .conf import settings
 
+handle_connection_failure = identity
+
+client_class_name = getattr(settings, 'CACHEOPS_CLIENT_CLASS', None)
+client_class = import_string(client_class_name) if client_class_name else redis.StrictRedis
+
+
+def ip(hostname):
```
Comment (nicwolff) [inline cacheops/redis.py]: Since the Redis settings are now dicts instead of URLs, we don't have to parse the hostnames out any more.

File: cacheops/redis.py [hunk 3]
```diff
@@ -1,20 +1,37 @@
...
-import random
+from funcy import decorator, identity, memoize
+import redis
 
 from .conf import settings
 
+handle_connection_failure = identity
+
+client_class_name = getattr(settings, 'CACHEOPS_CLIENT_CLASS', None)
+client_class = import_string(client_class_name) if client_class_name else redis.StrictRedis
```
Comment (nicwolff) [inline cacheops/redis.py]: The diff is hard to follow, but really all I've done is set the `client_class` here, either to one passed in by name as a setting or to the default `redis.StrictRedis` client.

File: cacheops/redis.py [hunk 4]
```diff
@@ -1,20 +1,37 @@
...
+
+def ip(hostname):
+    try:
+        return socket.gethostbyname(hostname)
+    except socket.gaierror as err:
+        warnings.warn('Hostname %s did not resolve because %s' % (hostname, err))
+        raise
+
+
 if settings.CACHEOPS_DEGRADE_ON_FAILURE:
```
Comment (nicwolff) [inline cacheops/redis.py]: If this setting is on, we want to handle connection failures. This code is back to pretty much how it was in cacheops 3.0.1 before we forked.

Comment (nicwolff) [inline cacheops/redis.py]: If this setting is on, we want to handle connection failures. This code is back to pretty much how it was in cacheops 3.0.1 before we forked.

File: cacheops/redis.py [hunk 5]
```diff
@@ -25,115 +42,74 @@ def handle_connection_failure(call):
...
+    client_class = SafeRedis
 
-def ip(url):
-    try:
-        return socket.gethostbyname(urlparse(url).hostname)
-    except socket.gaierror:
-        warnings.warn('Hostname in URL %s did not resolve' % url)
-        raise
 
+if settings.CACHEOPS_REDIS_REPLICAS:
```
Comment (nicwolff) [inline cacheops/redis.py:46]: And now we can put all our read-replica code in one place! If there are any replicas in the settings, we subclass the `client_class` to handle them.

File: cacheops/redis.py [hunk 6]
```diff
@@ -25,115 +42,74 @@ def handle_connection_failure(call):
...
 
-    redis_urls = settings.REDIS_REPLICAS
-    # the conf could be a list or string
-    # list would look like: ["redis://cache-001:6379/1", "redis://cache-002:6379/2"]
-    # string would be: "redis://cache-001:6379/1,redis://cache-002:6379/2"
-    if isinstance(redis_urls, six.string_types):
-        redis_urls = redis_urls.split(',')
-    else:
-        redis_urls = list(redis_urls)
+        read_clients = []
```
Comment (nicwolff) [inline cacheops/redis.py]: Now a class attribute instead of a global.

File: cacheops/redis.py [hunk 7]
```diff
@@ -25,115 +42,74 @@ def handle_connection_failure(call):
...
 
-    # Make Redis clients from all the URLs except the primary
-    new_read_clients = [redis.StrictRedis.from_url(u) for u in redis_urls if ip(u) != primary_ip]
+        @classmethod
+        def set_read_clients(cls):
+            primary_ip = ip(settings.CACHEOPS_REDIS['host'])
+            replicas = settings.CACHEOPS_REDIS_REPLICAS
+            replica_weight = settings.CACHEOPS_REPLICA_WEIGHT
```
Comment (nicwolff) [inline cacheops/redis.py]: We were using the non-cacheops settings `REDIS_MASTER` and `REDIS_REPLICAS`. Now we use the ones set up for this library, which are always `dict`s not URL strings.

File: cacheops/redis.py [hunk 8]
```diff
@@ -25,115 +42,74 @@ def handle_connection_failure(call):
...
+    class ReplicaProxyRedis(client_class):
 
-    redis_urls = settings.REDIS_REPLICAS
-    # the conf could be a list or string
-    # list would look like: ["redis://cache-001:6379/1", "redis://cache-002:6379/2"]
-    # string would be: "redis://cache-001:6379/1,redis://cache-002:6379/2"
-    if isinstance(redis_urls, six.string_types):
-        redis_urls = redis_urls.split(',')
-    else:
-        redis_urls = list(redis_urls)
```
Comment (nicwolff) [inline cacheops/redis.py:56]: We're always passing in a list now.

File: cacheops/redis.py [hunk 9]
```diff
@@ -25,115 +42,74 @@ def handle_connection_failure(call):
...
+        def execute_command(self, *args, **options):
+            """Handle failover of AWS elasticache."""
+            try:
+                return super(SafeRedis, self).execute_command(*args, **options)
+            except redis.ResponseError as e:
+                if "READONLY" not in e.message:
+                    raise
+                connection = self.connection_pool.get_connection(args[0], **options)
+                connection.disconnect()
+                self.read_clients = []
```
Comment (nicwolff) [inline cacheops/redis.py:94]: This should cause the next `.get()` call to reload the replica list, identifying the new primary by IP. Don't know if it'll work, but we are deploying this anyway since we need to start reading from the primary.

File: cacheops/redis.py [hunk 10]
```diff
@@ -25,115 +42,74 @@ def handle_connection_failure(call):
...
-### Lua script loader
+    client_class = ReplicaProxyRedis
 
-import re
-import os.path
+
+if isinstance(settings.CACHEOPS_REDIS, six.string_types):
+    redis_client = client_class.from_url(settings.CACHEOPS_REDIS)
+else:
+    redis_client = client_class(**settings.CACHEOPS_REDIS)
```
Comment (nicwolff) [inline cacheops/redis.py]: In theory we should still be handling Redis URLs as well as dicts – we can clean that up after we update our fork onto cacheops 4.

**Comment by nicwolff** [comment]
Rover passed tests in Travis when I installed this by SHA here https://travis-ci.com/HearstCorp/rover/builds/56449600

**Review by akarpodinis — APPROVED** [review]
I keep wanting to ask for unit tests, because the function design really wants it, then I remember this is a fork ☹️ .

**Comment by nicwolff** [comment]
Not doing them now, but I hope to contribute much of this back to the cacheops project, by which point it'll need unit tests.

File: cacheops/redis.py [hunk 11]
```diff
@@ -1,20 +1,31 @@
...
-from urlparse import urlparse
+import warnings
+from copy import copy
 
-from funcy import decorator, identity, memoize
 import redis
 from django.core.exceptions import ImproperlyConfigured
 from django.utils.module_loading import import_string
-import random
+from funcy import decorator, identity, memoize
```
Comment (nicwolff) [inline cacheops/redis.py:16]: Sorted the imports.

**Comment by nicwolff** [comment]
Dennis can't log into GitHub until T-Mobile works on 10, but he OKed this.

**Comment by nicwolff** [comment]
Passed QA on the Rover PR here https://github.com/HearstCorp/rover/pull/950#issuecomment-336268768
